api/v1beta1/types.go

```diff
@@ -324,6 +324,44 @@ type Image struct {
 	// Marketplace specifies an image to use from the Azure Marketplace
 	// +optional
 	Marketplace *AzureMarketplaceImage `json:"marketplace,omitempty"`
+
+	// CommunityGallery specifies an image to use from the Azure Community Gallery
+	// +optional
+	CommunityGallery *AzureCommunityGalleryImage `json:"communityGallery,omitempty"`
+}
+
+// AzureCommunityGalleryImage defines an image in the Azure Community Gallery to use for VM creation.
+type AzureCommunityGalleryImage struct {
```

Comment:
seems like there is a lot of overlap between community galleries and shared galleries (which have since been renamed to compute galleries), have you considered either reusing the existing type, creating a new one that applies to both and deprecating the old one, or extracting the common fields to a struct that can be included inline in both shared and community galleries that contains the gallery details?

Reply:
I guess it could make sense, I'll have a look.

Reply:
I'm personally not a fan of option 2 ("Deprecate both SharedGallery and Marketplace, move all fields into Image struct") because it puts a lot of burden on the user to figure out which fields are relevant for which image type. 

My suggestion was simply to avoid introducing a whole new type for Community galleries as they are logically the same as Azure Compute (private) galleries, previously known as Shared Image Galleries. In fact, according to docs, a Compute gallery can be mutated to become public after it exists (and vice versa): https://docs.microsoft.com/en-us/azure/virtual-machines/create-gallery?tabs=portal%2Ccli2#create-a-community-gallery-preview

+1 to option 1 as it is what best reflects the logical distinction of Azure Image types and allows us to better align with the official rename of Azure Compute Galleries. 

In order to share fields we could also do something like this: 

```
type Image struct {
        ID             *string                   `json:"id,omitempty"`
        SharedGallery  *AzureSharedGalleryImage  `json:"sharedGallery,omitempty"`
        Marketplace    *AzureMarketplaceImage    `json:"marketplace,omitempty"`
        ComputeGallery *AzureComputeGalleryImage `json:"computeGallery,omitempty"`
}

type AzureSharedGalleryImage struct {
        AzureComputeGalleryImage `json:",inline"`
        SubscriptionID *string `json:"subscriptionID"`
        ResourceGroup  *string `json:"resourceGroup"`
}

type AzureComputeGalleryImage struct {
        ImagePlan `json:",inline"`
        Gallery        string  `json:"gallery"`
        Name           string  `json:"name"`
        Version        string  `json:"version"`
}

type ImagePlan struct {
        Publisher      *string `json:"publisher,omitempty"`
        Offer          *string `json:"offer,omitempty"`
        SKU            *string `json:"sku,omitempty"`
}

type AzureMarketplaceImage struct {
        ImagePlan `json:",inline"`
        Version         *string `json:"version"`
        ThirdPartyImage bool    `json:"thirdPartyImage"`
}
```

Reply:
Before I jump in into implementing all crazy conversion functions, what do you think about the types below?

So from what I saw, plan information must consist of Publisher, Offer and SKU, not having any of this fields populated results in an API error (correct me if there are cases when this is not true). Given that, let's make `ImagePlan` have only non-pointer required fields and control the requirements of those fields one level above.

So for `AzureMarketplaceImage`, `ImagePlan` is required, while for `AzureSharedGalleryImage` and `AzureComputeGalleryImage` fields are optional. I wonder if kubebuilder is smart enough about `+optional` on inlined fields.

`Version` field remains duplicated across all types for the time being.

And regardless of the change, for Go types this will be a breaking change. For YAML I think types remains backward compatible.

```go
// Image defines information about the image to use for VM creation.
// There are three ways to specify an image: by ID, Marketplace Image, ComputeGallery or SharedImageGallery
// One of ID, SharedImage, ComputeGallery or Marketplace should be set.
type Image struct {
        // ID specifies an image to use by ID
        // +optional
        ID *string `json:"id,omitempty"`

        // SharedGallery specifies an image to use from an Azure Shared Image Gallery
        // Deprecated: use computeGallery instead.
        // +optional
        SharedGallery *AzureSharedGalleryImage `json:"sharedGallery,omitempty"`

        // Marketplace specifies an image to use from the Azure Marketplace
        // +optional
        Marketplace *AzureMarketplaceImage `json:"marketplace,omitempty"`

        // ComputeGallery specifies an image to use from the Azure Compute Gallery
        // +optional
        ComputeGallery *AzureComputeGalleryImage `json:"computeGallery,omitempty"`
}

// AzureComputeGalleryImage defines an image in the Azure Compute Gallery to use for VM creation.
type AzureComputeGalleryImage struct {
        // ImagePlan contains `Plan` information for API request when creating the VM/VMSS resource.
        // +optional
        *ImagePlan `json:",inline"`

        // Gallery specifies the name of the compute gallery that contains the image. If specified without
        // subscriptionID and resourceGroup, image assumes reference to a community gallery.
        // +kubebuilder:validation:MinLength=1
        Gallery string `json:"gallery"`
        // Name is the name of the image
        // +kubebuilder:validation:MinLength=1
        Name string `json:"name"`
        // Version specifies the version of the compute gallery image. The allowed formats
        // are Major.Minor.Build or 'latest'. Major, Minor, and Build are decimal numbers.
        // Specify 'latest' to use the latest version of an image available at deploy time.
        // Even if you use 'latest', the VM image will not automatically update after deploy
        // time even if a new version becomes available.
        // +kubebuilder:validation:MinLength=1
        Version string `json:"version"`

        // SubscriptionID is the identifier of the subscription that contains the compute gallery.
        // Must be specified together with resourceGroup. if not specified, image assumes reference to
        // a community gallery.
        // +optional
        SubscriptionID *string `json:"subscriptionID"`
        // ResourceGroup specifies the resource group containing the compute gallery. Must be specified
        // together with subscriptionID. If not specified, image assumes reference to a community gallery.
        // +optional
        ResourceGroup *string `json:"resourceGroup"`
}

// ImagePlan contains `Plan` information for API request when creating the VM/VMSS resource.
type ImagePlan struct {
        // Publisher is the name of the organization that created the image.
        // This value will be used to add a `Plan` in the API request when creating the VM/VMSS resource.
        // This is needed when the source image from which this SIG image was built requires the `Plan` to be used.
        // +kubebuilder:validation:MinLength=1
        Publisher string `json:"publisher,omitempty"`
        // Offer specifies the name of a group of related images created by the publisher.
        // For example, UbuntuServer, WindowsServer
        // This value will be used to add a `Plan` in the API request when creating the VM/VMSS resource.
        // This is needed when the source image from which this SIG image was built requires the `Plan` to be used.
        // +kubebuilder:validation:MinLength=1
        Offer string `json:"offer,omitempty"`
        // SKU specifies an instance of an offer, such as a major release of a distribution.
        // For example, 18.04-LTS, 2019-Datacenter
        // This value will be used to add a `Plan` in the API request when creating the VM/VMSS resource.
        // This is needed when the source image from which this SIG image was built requires the `Plan` to be used.
        // +kubebuilder:validation:MinLength=1
        SKU string `json:"sku,omitempty"`
}

// AzureSharedGalleryImage defines an image in a Shared Image Gallery to use for VM creation.
type AzureSharedGalleryImage struct {
        // ImagePlan contains `Plan` information for API request when creating the VM/VMSS resource.
        // +optional
        *ImagePlan `json:",inline"`

        // SubscriptionID is the identifier of the subscription that contains the shared image gallery
        // +kubebuilder:validation:MinLength=1
        SubscriptionID string `json:"subscriptionID"`
        // ResourceGroup specifies the resource group containing the shared image gallery
        // +kubebuilder:validation:MinLength=1
        ResourceGroup string `json:"resourceGroup"`
        // Gallery specifies the name of the shared image gallery that contains the image
        // +kubebuilder:validation:MinLength=1
        Gallery string `json:"gallery"`
        // Name is the name of the image
        // +kubebuilder:validation:MinLength=1
        Name string `json:"name"`
        // Version specifies the version of the marketplace image. The allowed formats
        // are Major.Minor.Build or 'latest'. Major, Minor, and Build are decimal numbers.
        // Specify 'latest' to use the latest version of an image available at deploy time.
        // Even if you use 'latest', the VM image will not automatically update after deploy
        // time even if a new version becomes available.
        // +kubebuilder:validation:MinLength=1
        Version string `json:"version"`
}

// AzureMarketplaceImage defines an image in the Azure Marketplace to use for VM creation.
type AzureMarketplaceImage struct {
        // ImagePlan contains `Plan` information for API request when creating the VM/VMSS resource.
        ImagePlan `json:",inline"`

        // Version specifies the version of an image sku. The allowed formats
        // are Major.Minor.Build or 'latest'. Major, Minor, and Build are decimal numbers.
        // Specify 'latest' to use the latest version of an image available at deploy time.
        // Even if you use 'latest', the VM image will not automatically update after deploy
        // time even if a new version becomes available.
        // +kubebuilder:validation:MinLength=1
        Version string `json:"version"`
        // ThirdPartyImage indicates the image is published by a third party publisher and a Plan
        // will be generated for it.
        // +kubebuilder:default=false
        // +optional
        ThirdPartyImage bool `json:"thirdPartyImage"`
}
```

Reply:
Honestly, I don't like existing plan fields definitions in `AzureSharedGalleryImage` either. They are currently defined as:
```go
	// +optional
	Publisher *string `json:"publisher,omitempty"`
	// +optional
	Offer *string `json:"offer,omitempty"`
	// +optional
	SKU *string `json:"sku,omitempty"`
```

Created #2286 around it.

I'd rather see those fields in a struct as they are currently defined in `AzureMarketplaceImage`:
```go
  // Publisher is the name of the organization that created the image
  // +kubebuilder:validation:MinLength=1
  Publisher string `json:"publisher"`
  // Offer specifies the name of a group of related images created by the publisher.
  // For example, UbuntuServer, WindowsServer
  // +kubebuilder:validation:MinLength=1
  Offer string `json:"offer"`
  // SKU specifies an instance of an offer, such as a major release of a distribution.
  // For example, 18.04-LTS, 2019-Datacenter
  // +kubebuilder:validation:MinLength=1
  SKU string `json:"sku"`
```

And in `AzureMarketplaceImage` they would be inlined, while in `AzureComputeGalleryImage`, they would be part of `+optional` `Plan` field. And we would deprecate `AzureSharedGalleryImage`. Thoughts?

Reply:
I can update `AzureMarketplaceImage`, on the other hand it could be done completely independently after introducing the right type.

Reply:
Done what proposed above. Also extracted compute update to #2288, so perhaps it can be merged sooner than this PR.

Reply:
Will do a full review soon, but this structure looks good to me 👍